# Run with --help or -h flag for help.
# Written 12/31/2018 by Fabio Marroni
suppressPackageStartupMessages({
  library(optparse)
})

option_list = list(
  make_option(c("-F", "--orig_fasta"), type="character", default="",help="Original trinity fasta file [default= %default]", metavar="character"), 
  make_option(c("-M", "--mod_fasta"), type="character", default="",help="Modified trinity fasta file (with description flanking name) [default= %default]", metavar="character"),
  make_option(c("-G", "--gtf_file"), type="character", default="", help="gtf file generated by stringtie (contains fasta names plus names assigned by stringtie) [default= %default]", metavar="character"),
  make_option(c("-D", "--DEseq_file"), type="character", default="", help="differential expression file (contains DE results and only names assigned by stringtie) [default= %default]", metavar="character"),
  make_option(c("-O", "--out"), type="character", default="", help="output file (contains DE results and assignment to species) [default= %default]", metavar="character")
)

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

#print("USAGE: $ 01_small_RNA.r -I input dir/ -O summarized output file")

if (is.null(opt$orig_fasta)) {
  stop("WARNING: No original fasta specified with '-F' flag.")
} else {  cat ("Original fasta is ", opt$orig_fasta, "\n")
  orig_fasta <- opt$orig_fasta  
  }

if (is.null(opt$mod_fasta)) {
  stop("WARNING: No modified fasta specified with '-M' flag.")
} else {  cat ("Modified fasta is ", opt$mod_fasta, "\n")
  mod_fasta <- opt$mod_fasta  
  }

if (is.null(opt$gtf_file)) {
  stop("WARNING: No gtf file specified with '-G' flag.")
} else {  cat ("Gtf file is ", opt$gtf_file, "\n")
  gtf_file <- opt$gtf_file  
  }

if (is.null(opt$DEseq_file)) {
  stop("WARNING: No DEseq result file specified with '-D' flag.")
} else {  cat ("DEseq result file is ", opt$DEseq_file, "\n")
  DEseq_file <- opt$DEseq_file  
  }


if (is.null(opt$out)) {
  stop("WARNING: No output file '-O' flag.")
} else {  cat ("Output file is ", opt$out, "\n")
  outfile <- opt$out  
  #setwd(wd_location)  
  }

add_descriptions<-function(orig_fasta, mod_fasta, gtf_file, DEseq_file, outfile) 
{
	library(seqinr)
    library(data.table)
	cat("Reading original fasta\n")	
	of<-read.fasta(orig_fasta)
	alltnames<-data.frame(names(of),stringsAsFactors=F)    
	setnames(alltnames,"trinity_id")
	cat("Reading modified fasta\n")	
	mf<-scan(mod_fasta,sep="\n",what="")
	mf<-mf[grep(">",mf)]
	mf<-gsub(">","",mf)	
	tnames<-unlist(lapply(strsplit(mf," "),"[",1))	
	cc<-lapply(strsplit(mf," "),"[",-1)	
	modtnames<-unlist(lapply(cc,paste,sep="_",collapse="_"))	
	mnames<-data.frame(tnames,modtnames,stringsAsFactors=F)
	setnames(mnames,c("trinity_id","taxonomy"))	
	cat("Reading gtf file\n")	
	gtf<-fread(gtf_file,data.table=F)
	gtf<-gtf[gtf$V3=="transcript",]	
	gtf<-gtf[!duplicated(gtf[,c(1,9)]),]	
	gtf<-gtf[,c(1,4,5,9)]
	gtf$V9<-unlist(lapply(strsplit(gtf$V9,"\""),"[",2))
	gtf<-gtf[!duplicated(gtf),]	
	setnames(gtf,c("trinity_id","start","end","gene_id"))	
	DE<-fread(DEseq_file,data.table=F,fill=TRUE)
	#In some instances we have one initial stupid Row.names column that we need to remove
	DE$Row.names<-NULL
	mynames<-names(DE)
	mynames<-mynames[1:(length(mynames)-1)]
	mynames<-c("gene_id",mynames)
	setnames(DE,mynames)
	cat("Merging full transcript list with the names mapped by kraken on the modified fasta\n")
	fullnames<-merge(alltnames,mnames,by="trinity_id",all=T)
	fullgtf<-merge(fullnames,gtf,by="trinity_id",all=T)	
	fullres<-merge(DE,fullgtf,by="gene_id",all.x=T,all.y=F,sort=F)	
	fullres<-fullres[!duplicated(cbind(fullres$gene_id,fullres$trinity_id)),]	
	write.table(fullres,outfile,sep="\t",quote=F,row.names=F)
}
add_descriptions(orig_fasta, mod_fasta, gtf_file, DEseq_file, outfile)
