#!/bin/bash

##############################################
#Variables you will need to change
###############################################
PROJDIR=/path/to/results  #Everything will be saved as a subfolder of this
SOFTDIR=/path/to/software #Assuming you have a common repository for software (only needed if the software packages I used are not in your path)
FUNCDIR=/path/to/scripts/and/functions #I suggest you create inside this folder a folder called "functions/limodorum" in which you will save all the R functions provided.
										# You also need to create a folder scripts/limodorum in which you need to put al the shell and python scripts
ANACDIR=/path/to/anaconda/bins #I needed it to quickly access my conda environments
KRAKBIN=/path/to/kraken/executables # Only needed if kraken is not in your path nor in SOFTDIR (I am a mess, and it was in its own folder!)
KDB=/path/to/kraken/db #No comments
KSCRIPT=/path/to/kraken/helper/scripts #
TAXKIT=/path/to/taxonkit/bin        #Only needed if taxonkit is not in your repository specified with SOFTDIR
##############################################
#End of variables you will need to change
###############################################

#Use my python script to change the name of the fasta sequences so that they also include the organism name. This will be used for future work, to discriminate between the orchid and the fungi.
python ${FUNCDIR}/scripts/limodorum/rename_seq.py --fasta_in ${PROJDIR}/full_trinity_assembly/Trinity.fasta --kraken_in ${PROJDIR}/kraken_nt/Trinity.kraken --fasta_output ${PROJDIR}/full_trinity_assembly/Trinity_nt.fasta


#Create a table with original name of transcript, the novel transcript name (generated by stringtie) and the description reporting some information regarding the species.
#Each possible annotation of the transcript is reported, so some duplicated entries are present. 
#I kept both to have the possibility of running blast on both the forms in case is more specific for classification.


module load it/lang/r/3.3
for ORG in AB plants
do
#New version for adding raw or normalized counts to DE results and then bring everything with us when adding GO and KEGG (in following scripts)
for count in norm raw
do
Rscript ${FUNCDIR}/functions/limodorum/01_merge_names_descriptions.r\
 -F ${PROJDIR}/full_trinity_assembly/Trinity.fasta\
 -M ${PROJDIR}/full_trinity_assembly/Trinity_nt.fasta\
 -G ${PROJDIR}/gtf/merged.gtf\
 -D ${PROJDIR}/DE_${ORG}/${ORG}_DE_${count}_2020.txt\
 -O ${PROJDIR}/DE_${ORG}/${ORG}_DE_${count}_2020_new.txt
done
done
